cmd/cloud-controller-manager/nodeipamcontroller.go

```diff
@@ -78,60 +78,56 @@ func startNodeIpamController(ccmConfig *cloudcontrollerconfig.CompletedConfig, n
-		if err != nil {
-			return nil, false, err
-		}
-		clusterCIDRs = cidrs
-
-		// failure: more than one cidr but they are not configured as dual stack
-		if len(clusterCIDRs) > 1 && !dualStack {
-			return nil, false, fmt.Errorf("len of ClusterCIDRs==%v and they are not configured as dual stack (at least one from each IPFamily", len(clusterCIDRs))
-		}
+	// failure: bad cidrs in config
+	cidrs, dualStack, err := processCIDRs(ccmConfig.ComponentConfig.KubeCloudShared.ClusterCIDR)
+	if err != nil {
+		return nil, false, err
+	}
+	clusterCIDRs = cidrs
 
-		// failure: more than cidrs is not allowed even with dual stack
-		if len(clusterCIDRs) > 2 {
-			return nil, false, fmt.Errorf("len of clusters is:%v > more than max allowed of 2", len(clusterCIDRs))
-		}
+	// failure: more than one cidr but they are not configured as dual stack
+	if len(clusterCIDRs) > 1 && !dualStack {
+		return nil, false, fmt.Errorf("len of ClusterCIDRs==%v and they are not configured as dual stack (at least one from each IPFamily", len(clusterCIDRs))
+	}
 
-		// service cidr processing
-		if len(strings.TrimSpace(nodeIPAMConfig.ServiceCIDR)) != 0 {
-			_, serviceCIDR, err = net.ParseCIDR(nodeIPAMConfig.ServiceCIDR)
-			if err != nil {
-				klog.Warningf("Unsuccessful parsing of service CIDR %v: %v", nodeIPAMConfig.ServiceCIDR, err)
-			}
-		}
+	// failure: more than 2 cidrs is not allowed even with dual stack
+	if len(clusterCIDRs) > 2 {
+		return nil, false, fmt.Errorf("len of clusters cidrs is:%v > more than max allowed of 2", len(clusterCIDRs))
+	}
 
-		if len(strings.TrimSpace(nodeIPAMConfig.SecondaryServiceCIDR)) != 0 {
-			_, secondaryServiceCIDR, err = net.ParseCIDR(nodeIPAMConfig.SecondaryServiceCIDR)
-			if err != nil {
-				klog.Warningf("Unsuccessful parsing of service CIDR %v: %v", nodeIPAMConfig.SecondaryServiceCIDR, err)
-			}
+	// service cidr processing
+	if len(strings.TrimSpace(nodeIPAMConfig.ServiceCIDR)) != 0 {
+		_, serviceCIDR, err = net.ParseCIDR(nodeIPAMConfig.ServiceCIDR)
+		if err != nil {
+			klog.Warningf("Unsuccessful parsing of service CIDR %v: %v", nodeIPAMConfig.ServiceCIDR, err)
 		}
+	}
 
-		// the following checks are triggered if both serviceCIDR and secondaryServiceCIDR are provided
-		if serviceCIDR != nil && secondaryServiceCIDR != nil {
-			// should be dual stack (from different IPFamilies)
-			dualstackServiceCIDR, err := netutils.IsDualStackCIDRs([]*net.IPNet{serviceCIDR, secondaryServiceCIDR})
-			if err != nil {
-				return nil, false, fmt.Errorf("failed to perform dualstack check on serviceCIDR and secondaryServiceCIDR error:%v", err)
-			}
-			if !dualstackServiceCIDR {
-				return nil, false, fmt.Errorf("serviceCIDR and secondaryServiceCIDR are not dualstack (from different IPfamiles)")
-			}
+	if len(strings.TrimSpace(nodeIPAMConfig.SecondaryServiceCIDR)) != 0 {
+		_, secondaryServiceCIDR, err = net.ParseCIDR(nodeIPAMConfig.SecondaryServiceCIDR)
+		if err != nil {
```

Comment:
Does this logic need to change once K8s supports multiple service CIDRs?

Reply:
Thanks for this.

Speaking about GKE, as far as I understand, the primary and secondary service CIDRs are configured on the IPAM controller through parameters for the cloud-controller-manager component, and currently there will never be more than 2 service CIDRs configured.

You are referring to something different, right? What are the timelines for Kubernetes to support more CIDRs? I think we need to adjust the code if this is going to happen, but maybe not as part of this PR? WDYT?

Reply:
I am referring to this: https://github.com/kubernetes/kubernetes/pull/116516 which is in alpha in 1.29

Reply:
Yes I think we can resolve this. The default config will never have multiple service CIDRs.